support/mesh/mesh_channel_core/src/mpsc.rs

```diff
@@ -0,0 +1,971 @@
+
+impl<T> Default for Receiver<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+struct ReceiverCore {
+    queue: ReceiverQueue,
+    ports: PortHandlerList,
+    terminated: bool,
+}
+
+#[derive(Debug)]
+struct ReceiverQueue(Arc<Queue>);
+
+impl Drop for ReceiverQueue {
+    fn drop(&mut self) {
+        let mut local = self.0.local.lock();
+        local.receiver_gone = true;
+        let _waker = std::mem::take(&mut local.waker);
+        local.messages.clear_and_shrink();
+        let _ports = std::mem::take(&mut local.ports);
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Creates a new receiver with no senders.
+    ///
+    /// Receives will fail with [`RecvError::Closed`] until [`Self::sender`] is
+    /// called.
+    pub fn new() -> Self {
+        Self(
+            ReceiverCore::new(const { &ElementVtable::new::<T>() }),
+            PhantomData,
+        )
+    }
+
+    /// Consumes and returns the next message, waiting until one is available.
+    ///
+    /// Returns immediately when the channel is closed or failed.
+    ///
+    /// ```rust
+    /// # use mesh_channel_core::*;
+    /// # futures::executor::block_on(async {
+    /// let (send, mut recv) = channel();
+    /// send.send(5u32);
+    /// drop(send);
+    /// assert_eq!(recv.recv().await.unwrap(), 5);
+    /// assert!(matches!(recv.recv().await.unwrap_err(), RecvError::Closed));
+    /// # });
+    /// ```
+    pub fn recv(&mut self) -> Recv<'_, T> {
+        Recv(self)
+    }
+
+    /// Consumes and returns the next message, if there is one.
+    ///
+    /// Otherwise, returns whether the channel is empty, closed, or failed.
+    ///
+    /// ```rust
+    /// # use mesh_channel_core::*;
+    /// let (send, mut recv) = channel();
+    /// send.send(5u32);
+    /// drop(send);
+    /// assert_eq!(recv.try_recv().unwrap(), 5);
+    /// assert!(matches!(recv.try_recv().unwrap_err(), TryRecvError::Closed));
+    /// ```
+    pub fn try_recv(&mut self) -> Result<T, TryRecvError> {
+        let mut v = MaybeUninit::<T>::uninit();
+        // SAFETY: `v` is a valid uninitialized `T`.
+        let r = unsafe { self.0.try_poll_recv(None, MessagePtr::new(&mut v)) };
+        match r {
+            Poll::Ready(Ok(())) => {
+                // SAFETY: `try_poll_recv` guarantees `v` is now initialized.
+                Ok(unsafe { v.assume_init() })
+            }
+            Poll::Ready(Err(RecvError::Closed)) => Err(TryRecvError::Closed),
+            Poll::Ready(Err(RecvError::Error(e))) => Err(TryRecvError::Error(e)),
+            Poll::Pending => Err(TryRecvError::Empty),
+        }
+    }
+
+    /// Polls for the next message.
+    ///
+    /// If one is available, consumes and returns it. If the
+    /// channel is closed or failed, fails. Otherwise, registers the current task to wake
+    /// when a message is available or the channel is closed or fails.
+    pub fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Result<T, RecvError>> {
```

Comment:
is this a useful method to have on the `Receiver` directly, vs. pushing this onto the `Recv` type directly?

Reply:
You're right, technically callers today don't even need this; they can just use `.recv().poll()`, since `Recv` is stateless.

This may change in the future, though. E.g., we could make `Recv` `!Unpin` and have it provide a slot to put the message in directly, so that if the receiver is waiting, the sender can skip the queue and write the message directly to the stack. This would also allow us to avoid taking the lock on subsequent polls, which would make some of the `select!` business cheaper.

But in this model, you might still want this (now less efficient, comparatively) method on `Receiver` for contexts where you can't pin, for whatever reason.

Anyway, existing callers depend on this so I'll leave it for now.